Consolidate onBuild tests + verify imagesPath for multiple operating systems

gshel opened the pull request:
# Description

Consolidated and added new tests to demonstrate bug where Windows imagesPath contains url-encoded `\`, caused by path.join() using os specific separator which is escaped `\` for windows

## Issue Ticket Number
`#100`

Fixes 

## Type of change

<!-- Please select all options that are applicable. -->

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update


# Checklist

<!-- These must all be followed and checked. -->

- [x] I have followed the contributing guidelines of this project as mentioned in [CONTRIBUTING.md](/CONTRIBUTING.md)
- [x] I have created an [issue](https://github.com/colbyfayock/netlify-plugin-cloudinary/issues) ticket for this PR
- [x] I have checked to ensure there aren't other open [Pull Requests](https://github.com/colbyfayock/netlify-plugin-cloudinary/pulls) for the same update/change?
- [x] I have performed a self-review of my own code
- [x] I have run tests locally to ensure they all pass
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes needed to the documentation

gshel commented on netlify-plugin-cloudinary/tests/on-build.test.js:
```diff
@@ -1,8 +1,20 @@
+    url: 'https://deploy-preview-1234--netlify-plugin-cloudinary.netlify.app'
+  }
+]
+
```
For deploy-preview, I noticed you were using the DEPLOY_PRIME_URL instead. I'd be happy to readd that, just let me know~

gshel commented:
> hey @gshel changes generally look good! looks like it's running into a little issue though
> 
> are you able to see the logs? https://github.com/cloudinary-community/netlify-plugin-cloudinary/actions/runs/6885078686/job/18751733492?pr=101

I can, thanks Colby! Looks like I've got some work to do on the win32 stuff.

I see that there is a posix test failing when `imagesPath = '/images/nest'`. I did some testing locally and it looks like if imagesPath is passed in as a string, it will always end up as `/images` regardless of the user input. Is this expected?

colbyfayock commented:
definitely not expected! that may be a regression then? as it did work at one point. i dont currently have a test for that it seems

gshel commented:
I crack myself up. I wrote tests to catch issues caused by using `path.sep` and then proceeded to use `path.sep` IN the tests 😶‍🌫️

gshel commented:
Okey doke I think it should be ready to pass all the tests now 🤞 

I fixed the issue where `imagesPath` would ignore the user `input['imagesPath']` if it was a string, and instead always use `/images`

I also added tests to make sure that imagesPath will:

1. default to `/images` if input.imagesPath is undefined.
e.g. `input['imagesPath'] === undefined` 
-> `imagesPath = '/images'`
2. use full value of input.imagesPath if it is defined as a string 
e.g. `input['imagesPath'] === '/babygoats/here'` 
-> `imagesPath = '/babygoats/here'`

There's quite a bit in this one PR; if you'd prefer I'd be happy to split it into separate PRs.

Thanks @colbyfayock !

colbyfayock commented on netlify-plugin-cloudinary/src/index.ts:
```diff
@@ -134,9 +134,7 @@ export async function onBuild({
-    imagesPath = CLOUDINARY_ASSET_DIRECTORIES.find(
-      ({ inputKey }) => inputKey === 'imagesPath',
-    )?.path,
+    imagesPath = inputs.imagesPath || '/images',
```
looking at this again - the intent for using `CLOUDINARY_ASSET_DIRECTORIES` was so that i could define the different configurations for different asset types in one location

for instance, in the future there could be a second object in `CLOUDINARY_ASSET_DIRECTORIES` with a name of `videos`

im kind of surprised the default wasn't working properly as the code seems to run fine

<img width="464" alt="image" src="https://github.com/cloudinary-community/netlify-plugin-cloudinary/assets/1045274/875ff21c-60bf-4131-ad8b-854d219a62ec">

can we keep the default value but use the `inputs` first if available like you have?

gshel replied:
> can we keep the default value but use the inputs first if available like you have?

I'll look into this now~

gshel replied:
✅

colbyfayock commented on netlify-plugin-cloudinary/src/index.ts:
```diff
@@ -277,6 +275,7 @@ export async function onBuild({
           }
 
           mediaPaths.forEach(async mediaPath => {
+            mediaPath = mediaPath.split(path.win32.sep).join(path.posix.sep);
```
just so im totally understanding what's happening here, basically, in the event that someone is on a windows machine, it will find the separator, then reconstruct it with posix separators, so that the rest of the code works as expected, but if its not windows, it basically wont do anything since it won't be there

and windows separators are `\`?

right?

colbyfayock commented:
just one more question on the default value update!

colbyfayock commented:
thanks @gshel!
